File: exercism/javascript/bob/bob.js
class Bob {
  hey(message) {
    message = message.trim();
    let regexp_normal = /\w$|[.!]$/;
    let regexp_q = /[?]$/;
    let regexp_loud = /[A-Z]{4,}|[A-Z]{2,}[!]$/;

    if (regexp_loud.test(message) && regexp_q.test(message)) {
      return "Calm down, I know what I\'m doing!";
    } else if (regexp_loud.test(message)) {
      return "Whoa, chill out!";
    } else if (regexp_q.test(message)) {
      return "Sure.";
    } else if (regexp_normal.test(message)) {
      return "Whatever.";
    } else {
      return "Fine. Be that way!";
    }
  };
};

export default Bob;

File: exercism/ruby/complete/bob/bob.rb
class Bob
  @@acronyms = { OK: true, DMV: true }

  def self.hey(string)
    string.strip!
    if string == ''
      "Fine. Be that way!"
    elsif loud?(string) && string[-1] != '?'
      "Whoa, chill out!"
    elsif loud?(string) && string[-1] == '?'
      "Calm down, I know what I'm doing!"
    elsif string[-1] == '?'
      "Sure."
    else
      "Whatever."
    end
  end

  private

  def self.loud?(string)
    string.gsub(/\W/, ' ').split.any? do |word|
      unless @@acronyms[word.to_sym] == true || word.length < 2
        word.chars.all?{ |letter| /[A-Z]/.match(letter) }
      end
    end
  end
end
